File: practicasros_ws/src/practica1/practica1/nodo1.py
import rclpy  # Importar la librería de ROS
from rclpy.node import Node  # Importar la clase Node
from std_msgs.msg import Int8  # Cambiado a Int8 para valores entre -128 y 127
import random

class Nodo1(Node):  # Crear la clase Nodo1 que hereda de Node
    def __init__(self):  # Constructor
        super().__init__('nodo1')  # Llamar al constructor de la clase padre
        self.publisher_ = self.create_publisher(Int8, 'numbers', 10)  # Publicador con Int8
        timer_period = 1.0  # Periodo del temporizador en segundos
        self.timer = self.create_timer(timer_period, self.timer_callback)  # Temporizador

    def timer_callback(self):  # Función que se llama cada vez que el temporizador se activa
        msg = Int8()  # Crear un mensaje de tipo Int8 (-128 a 127)
        msg.data = random.randint(-128, 127)  # ✅ Generar número aleatorio en el rango correcto
        self.publisher_.publish(msg)  # Publicar el mensaje
        self.get_logger().info(f'Publicando: {msg.data}')  # Imprimir en pantalla el mensaje publicado

def main(args=None):  # Función principal
    rclpy.init(args=args)  # Inicializar ROS
    nodo1 = Nodo1()  # Crear un objeto de la clase Nodo1
    rclpy.spin(nodo1)  # Mantener el programa en ejecución
    nodo1.destroy_node()  # Destruir el nodo
    rclpy.shutdown()  # Apagar ROS

if __name__ == '__main__':  # Si el script es el principal
    main()  # Llamar a la función principal

File: practicasros_ws/src/practica1_cpp/src/nodo1.cpp
#include <chrono>  // Para manejar el tiempo (1s)
#include <rclcpp/rclcpp.hpp>  // Librería de ROS 2 para C++
#include <functional>  // Para usar std::bind
#include <std_msgs/msg/int8.hpp>  // Tipo de mensaje a publicar (Int8)
#include <memory>  // Para usar std::shared_ptr
#include <random>  // Para generar números aleatorios

using namespace std::chrono_literals;  // Para usar "1s"

class Nodo1 : public rclcpp::Node {
public:
    Nodo1() : Node("nodo1") {
        // Crear publicador en el tópico "random_numbers"
        publisher_ = this->create_publisher<std_msgs::msg::Int8>("num", 10);

        // Configurar temporizador para publicar cada 1s
        timer_ = this->create_wall_timer(1s, std::bind(&Nodo1::publicar, this));

        RCLCPP_INFO(this->get_logger(), "Nodo1 iniciado - Publicando números aleatorios de 8 bits.");
    }

private:
    void publicar() {
        // Generar número aleatorio entre -128 y 127 (8 bits con signo)
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_int_distribution<int8_t> dist(-128, 127);

        std_msgs::msg::Int8 msg;
        msg.data = dist(gen);

        // Publicar el número aleatorio
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Publicado: %d", msg.data);
    }

    rclcpp::Publisher<std_msgs::msg::Int8>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Nodo1>());
    rclcpp::shutdown();
    return 0;
}